util/pkg/vfs/s3context.go

```diff
@@ -61,6 +64,50 @@ func (s *S3Context) getClient(region string) (*s3.S3, error) {
 	return s3Client, nil
 }
 
+func (s *S3Context) getEndpointClient(region string) (*s3.S3, error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s3Client := s.clients[region]
+	if s3Client == nil {
+		Endpoint := os.Getenv("S3_ENDPOINT")
+		if Endpoint == "" {
+			return nil, fmt.Errorf("S3_ENDPOINT is required")
+		}
+		AccessKeyID := os.Getenv("S3_ACCESS_KEY_ID")
+		if AccessKeyID == "" {
+			return nil, fmt.Errorf("S3_ACCESS_KEY_ID cannot be empty when S3_ENDPOINT is not empty")
+		}
+		SecretAccessKey := os.Getenv("S3_SECRET_ACCESS_KEY")
+		if SecretAccessKey == "" {
+			return nil, fmt.Errorf("S3_SECRET_ACCESS_KEY cannot be empty when S3_ENDPOINT is not empty")
+		}
+
+		s3Config := &aws.Config{
+			Credentials:      credentials.NewStaticCredentials(AccessKeyID, SecretAccessKey, ""),
+			Endpoint:         aws.String(Endpoint),
+			Region:           aws.String(region),
+			DisableSSL:       aws.Bool(true),
+			S3ForcePathStyle: aws.Bool(true),
+		}
+
+		session := session.New()
+		s3Client = s3.New(session, s3Config)
+		s.clients[region] = s3Client
+	}
+
+	return s3Client, nil
+}
+
+func (s *S3Context) setClientFunc() {
+	Endpoint := os.Getenv("S3_ENDPOINT")
+	if Endpoint == "" {
+		s.getClient = s.getDefaultClient
+	} else {
+		s.getClient = s.getEndpointClient
```

Comment:
It will save a lot of head-scratching later on, when someone accidentally has exported S3_ENDPOINT for whatever reason, to add `glog.Infof("Found S3_ENDPOINT=%q, using as non-AWS S3 backend")` here :-)